source/clear-linux/tutorials/fmv.rst

```diff
@@ -0,0 +1,268 @@
+                a[i] = b[i] + c[i];
+            }
+        }
+    }
+
+
+    int main(){
+        foo();
+        return 0;
+    }
+
+Save the example code as :file:`example.c` and build with the
+following flags:
+
+.. code-block:: bash
+
+        gcc -O3  -fopt-info-vec  example.c -o example
+
+The build generates the following output:
+
+.. code-block:: console
+
+    example.c:11:9: note: loop vectorized
+    example.c:11:9: note: loop vectorized
+
+The output shows that line 11 is a good candidate for vectorization:
+
+.. code-block:: c
+
+    for (i=0; i<256; i++){
+        a[i] = b[i] + c[i];
+
+Generate the FMV patch
+**********************
+
+To generate the FMV patch with the `make-fmv-patch`_ project, we
+must clone the project and generate a log file with the loop vectorized
+information:
+
+.. code-block:: bash
+
+        git clone https://github.com/clearlinux/make-fmv-patch.git
+        gcc -O3  -fopt-info-vec  example.c -o example &> log
+
+To generate the patch files, execute:
+
+.. code-block:: bash
+
+        perl ./make-fmv-patch/make-fmv-patch.pl log .
+
+The make-fmv-patch.pl take two arguments: <buildlog> and <sourcecode>. Replace with the proper values and execute:
+
+.. code-block:: bash
+
+        perl make-fmv-patch.pl <buildlog> <sourcecode>
+
+The command generates the following :file:`example.c.patch` patch:
+
+.. code-block:: console
+
+    --- ./example.c 2017-09-27 16:05:42.279505430 +0000
+    +++ ./example.c~    2017-09-27 16:19:11.691544026 +0000
+    @@ -5,6 +5,7 @@
+
+     int a[256], b[256], c[256];
+
+    +__attribute__((target_clones("avx2","arch=atom","default")))
+     void foo(){
+         int i,x;
+         for (x=0; x<MAX; x++){
+
+The `make-fmv-patch` is recommended to add the attribute generating the
+target clones on the function foo. Thus, we can have the following code:
+
+.. code-block:: c
+
+    #include <stdio.h>
+    #include <stdlib.h>
+    #include <sys/time.h>
+    #define MAX 1000000
+
+    int a[256], b[256], c[256];
+
+    __attribute__((target_clones("avx2","arch=atom","default")))
+    void foo(){
+        int i,x;
+        for (x=0; x<MAX; x++){
+            for (i=0; i<256; i++){
+                a[i] = b[i] + c[i];
+            }
+        }
+    }
+
+
+    int main(){
+        foo();
+        return 0;
+    }
+
+By changing the value of the `$avx2` variable, we can change the target clones when adding the patches or in the make-fmv-patch.pl script:
+
+.. code-block:: perl
+
+    my $avx2 = '__attribute__((target_clones("avx2","arch=atom","default")))'."\n";
+
+Compile the code again with FMV and add the option to analyze the `objdump`:
+
+.. code-block:: bash
+
+    gcc -O3 example.c -o example -g
+    objdump -S example | less
+
+You can see the multiple clones of the foo function:
+
+.. code-block:: console
+
+    foo
+    foo.avx2.0
+    foo.arch_atom.1
+
+The cloned functions use AVX2 registers and vectorized instructions. You can verify this with:
```

Comment:
Please use active voice. To verify, enter the following commands:

Reply:
I liked your consistent version much better and I've changed it. However, "You can ..." is active voice. ;)